Fixes #24640 - Cast all lookup keys and values

tbrisker opened the pull request:
db/migrate/20170112175131_migrate_template_to_parameters_macros.rb
caused all LookupKey default values and LookupValue values to be saved
as strings, because it updated the attributes without calling callbacks,
causing cast_default_value or cast_value callbacks to be ignored.
This causes any values not of string type to break.

When saving the LookupKey, all of its lookup values are also validated,
casted and saved, so we don't need to save them separately.



<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

tbrisker committed:
Fixes #24640 - Cast all lookup keys and values

db/migrate/20170112175131_migrate_template_to_parameters_macros.rb
caused all LookupKey default values and LookupValue values to be saved
as strings, because it updated the attributes without calling callbacks,
causing cast_default_value or cast_value callbacks to be ignored.
This causes any values not of string type to break.

ares commented:
[test foreman]

ares reviewed (changes requested):
It seems like tests failures are unrelated but broken on some webpack issue

```
WARNING: `rake webpack:compile` failed to run. This is only important if running integration tests. (cause: Command failed with status (1): [/var/lib/workspace/workspace/test_develop_...])
The Apipie cache is turned off. Enable it and run apipie:cache rake task to speed up API calls.
rake aborted!
ActiveSupport::DeprecationException: DEPRECATION WARNING: Static attributes will be removed in FactoryBot 5.0. Please use dynamic
attributes instead by wrapping the attribute value in a block:
```

I reproduced the problem and I can confirm this fixes the issue. Though if lookup key or lookup value is invalid, this migration will fail to fix the value. Saving models using active record can cause problems in future, e.g. when new validation on newly added attribute is added. When this migration runs on clean setup, it will fail in that scenario, since the field does not exist at this point.

We should call the callback explicitly and save it with validation skipping.

tbrisker commented:
@ares regarding invalid values - casting will fail anyways if the value is invalid, and we have no way of telling how to fix it. It will not fail the migration however, since we save each key separately it will only rollback the specific key (I have an invalid lookup value locally that doesn't get updated). If the value is invalid, calling the casting function will also fail for the same reason.
regarding adding new validations i'm not too concerned since we'll identify this at the PR stage since it will fail migrations.
However, looking at all the validations currently present, there are quite a few that aren't related to this that would be worth skipping, and we also handle the casting a bit differently for the different types of LKs. updated accordingly.

ares reviewed (commented):
That looks good now, I haven't tested yet. Does it already incorporate the fix for YAML values discussed on discourse? Or is that a separate issue?

tbrisker commented:
@ares it still doesn't fix the yaml issue (which i think might also be present for json type)

tbrisker commented:
Updated the PR to use safemode eval to parse all the values that may have been broken and directly update them in the database. This should now work for all key types and removed several cases that don't need to be cast (such as non-string values or omitted values)

tbrisker committed:
Refs #24640 - remove extra `()`

ares commented:
Thanks @tbrisker, the first commit passed and I tested it works as expected. Failures were unrelated. We won't wait for tests for the second commit as I also tested manually it works and risks are 0.

Merging this important fix, thanks for fixing regression I introduced once again.

tbrisker commented:
Cherry-picked to 1.19 as 635ca0e6fbd28c05ef0e54186b617e6601915f86
